[Build] Build scripts silently ignores non-accessible DLLs

Description
===========
<!--
A short one or two line description of the error
-->

When looking for available DLLs, the build scripts treats all access errors like files that don't exist:

https://github.com/arrayfire/arrayfire-rust/blob/7cfde6b827f2976767f4e5dc6b8913fe3ef0140e/build.rs#L71-L76

If the current user has no read permissions on the `af.dll`, it is not added to the `backends` list:

https://github.com/arrayfire/arrayfire-rust/blob/7cfde6b827f2976767f4e5dc6b8913fe3ef0140e/build.rs#L419-L424

As a result, no arrayfire library is passed to the linker, which then throws a lot of errors because for the various `af_*` functions (see below).

The arrayfire build script should really trigger an error or panic if no backends are found instead of letting the users run into linker errors. Ideally, the error message would also mention that some files were found, but were not be accessible.

Build Environment
-----------------
Compiler version: rustc 1.57.0 (f1edd0429 2021-11-29)
Operating system: Windows 10
Build environment: (described above)

Error Log
---------
<!-- Output of the error log. -->

Lots of errors like these:
```
error LNK2019: unresolved external symbol af_retain_array referenced in function _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h3e2e43e5f6399194E
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -73,6 +73,8 @@ jobs:
     clippy:
         name: Clippy Lints
         runs-on: ubuntu-18.04
+        env:
+            AF_VER: 3.8.0
         steps:
             - name: Checkout Repository
               uses: actions/checkout@master
@@ -83,7 +85,29 @@ jobs:
                   toolchain: stable
                   override: true
                   components: clippy
+
+            - name: Cache ArrayFire
+              uses: actions/cache@v1
+              id: arrayfire
+              with:
+                  path: afbin
+                  key: ${{ runner.os }}-af-${{ env.AF_VER }}
+
+            - name: Download ArrayFire
+              # Only download and cache arrayfire if already not found
+              if: steps.arrayfire.outputs.cache-hit != 'true'
+              run: |
+                  wget --quiet http://arrayfire.s3.amazonaws.com/${AF_VER}/ArrayFire-v${AF_VER}_Linux_x86_64.sh
+                  chmod +x ./ArrayFire-v${AF_VER}_Linux_x86_64.sh
+                  mkdir afbin
+                  ./ArrayFire-v${AF_VER}_Linux_x86_64.sh --skip-license --exclude-subdir --prefix=./afbin
+                  rm ./afbin/lib64/libcu*.so*
+                  rm ./afbin/lib64/libafcuda*.so*
+                  rm ./ArrayFire-v${AF_VER}_Linux_x86_64.sh
+
             - name: Run clippy tool
+              env:
+                AF_PATH: ${{ github.workspace }}/afbin
               uses: actions-rs/cargo@v1
               with:
                   command: clippy

--- a/build.rs
+++ b/build.rs
@@ -449,6 +449,11 @@ fn main() {
     }
 
     let (backends, backend_dirs) = blob_backends(&conf, &build_dir);
+
+    if backends.is_empty() {
+       fail("no arrayfire backends found");
+    }
+
     for backend in backends.iter() {
         println!("cargo:rustc-link-lib=dylib={}", backend);
     }

--- a/build.rs
+++ b/build.rs
@@ -292,13 +292,9 @@ fn blob_backends(conf: &Config, build_dir: &std::path::Path) -> (Vec<String>, Ve
         let afpath = match env::var("AF_PATH") {
             Ok(af_path) => PathBuf::from(&af_path),
             Err(_) => {
-                println!(
-                    "WARNING! USE_LIB is defined,
-                          but AF_PATH is not found,"
-                );
-                println!(
-                    "Trying to find libraries from
-                          known default locations"
+                eprintln!(
+                    "WARNING: USE_LIB is defined, but AF_PATH is not found. Trying to find \
+                    libraries from known default locations."
                 );
                 if cfg!(target_os = "windows") {
                     PathBuf::from("C:\\Program Files\\ArrayFire\\v3\\")

--- a/build.rs
+++ b/build.rs
@@ -71,7 +71,7 @@ fn dir_exists(location: &str) -> bool {
                 eprintln!("WARNING: failed to access `{}`: {}", location, err);
             }
             false
-        },
+        }
     }
 }
 
@@ -83,7 +83,7 @@ fn file_exists(location: &str) -> bool {
                 eprintln!("WARNING: failed to access `{}`: {}", location, err);
             }
             false
-        },
+        }
     }
 }
 
@@ -458,7 +458,7 @@ fn main() {
     let (backends, backend_dirs) = blob_backends(&conf, &build_dir);
 
     if backends.is_empty() {
-       fail("no arrayfire backends found");
+        fail("no arrayfire backends found");
     }
 
     for backend in backends.iter() {

--- a/build.rs
+++ b/build.rs
@@ -66,14 +66,24 @@ fn fail(msg: &str) -> ! {
 fn dir_exists(location: &str) -> bool {
     match fs::metadata(location) {
         Ok(f) => f.is_dir(),
-        Err(_) => false,
+        Err(err) => {
+            if err.kind() != ErrorKind::NotFound {
+                eprintln!("WARNING: failed to access `{}`: {}", location, err);
+            }
+            false
+        },
     }
 }
 
 fn file_exists(location: &str) -> bool {
     match fs::metadata(location) {
         Ok(f) => f.is_file(),
-        Err(_) => false,
+        Err(err) => {
+            if err.kind() != ErrorKind::NotFound {
+                eprintln!("WARNING: failed to access `{}`: {}", location, err);
+            }
+            false
+        },
     }
 }
 

--- a/build.rs
+++ b/build.rs
@@ -58,8 +58,9 @@ struct Config {
     win_vs_toolset: String,
 }
 
-fn fail(s: &str) -> ! {
-    panic!("\n{}\n\nbuild script failed, must exit now", s)
+fn fail(msg: &str) -> ! {
+    eprintln!("ERROR: {}", msg);
+    std::process::exit(1);
 }
 
 fn dir_exists(location: &str) -> bool {

